lib/SDIR/Translate/translateToSDFG.cpp

```diff
@@ -0,0 +1,1597 @@
+  jemit.printKVPair("__return", "null", /*stringify=*/false);
+  jemit.endObject(); // out_connectors
+
+  jemit.startNamedObject("sdfg");
+
+  if (printSDFGNode(op, jemit).failed())
+    return failure();
+
+  jemit.endObject(); // sdfg
+  jemit.endObject(); // attributes
+
+  jemit.endObject();
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// StateNode
+//===----------------------------------------------------------------------===//
+
+LogicalResult translateStateToSDFG(StateNode &op, JsonEmitter &jemit) {
+  jemit.startObject();
+  jemit.printKVPair("type", "SDFGState");
+  jemit.printKVPair("label", op.sym_name());
+  jemit.printKVPair("id", op.ID(), /*stringify=*/false);
+
+  jemit.startNamedObject("attributes");
+  jemit.printAttributes(op->getAttrs(), /*elidedAttrs=*/{"ID", "sym_name"});
+  if (!containsAttr(*op, "instrument"))
+    jemit.printKVPair("instrument", "No_Instrumentation");
+
+  jemit.endObject(); // attributes
+
+  jemit.startNamedList("nodes");
+
+  unsigned nodeID = 0;
+  for (Operation &oper : op.body().getOps()) {
+    if (TaskletNode tasklet = dyn_cast<TaskletNode>(oper)) {
+      tasklet.setID(nodeID++);
+      if (translateTaskletToSDFG(tasklet, jemit).failed())
+        return failure();
+    }
+
+    if (SDFGNode sdfg = dyn_cast<SDFGNode>(oper)) {
+      sdfg.setID(nodeID++);
+      if (translateSDFGToSDFG(sdfg, jemit).failed())
+        return failure();
+    }
+
+    if (GetAccessOp acc = dyn_cast<GetAccessOp>(oper)) {
+      acc.setID(nodeID++);
+      if (translateGetAccessToSDFG(acc, jemit).failed())
+        return failure();
+    }
+
+    if (MapNode map = dyn_cast<MapNode>(oper)) {
+      map.setEntryID(nodeID++);
+      map.setExitID(nodeID++);
+      if (translateMapToSDFG(map, jemit).failed())
+        return failure();
+    }
+
+    if (ConsumeNode consume = dyn_cast<ConsumeNode>(oper)) {
+      consume.setEntryID(nodeID++);
+      consume.setExitID(nodeID++);
+      if (translateConsumeToSDFG(consume, jemit).failed())
+        return failure();
+    }
+  }
+
+  jemit.endList(); // nodes
+
+  jemit.startNamedList("edges");
+
+  for (Operation &oper : op.body().getOps()) {
+    if (CopyOp edge = dyn_cast<CopyOp>(oper))
+      if (translateToSDFG(*edge, jemit).failed())
+        return failure();
+
+    if (StoreOp edge = dyn_cast<StoreOp>(oper))
+      if (translateToSDFG(*edge, jemit).failed())
+        return failure();
+
+    if (sdir::CallOp edge = dyn_cast<sdir::CallOp>(oper))
+      if (translateToSDFG(*edge, jemit).failed())
+        return failure();
+  }
+
+  jemit.endList(); // edges
+
+  jemit.endObject();
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// TaskletNode
+//===----------------------------------------------------------------------===//
+
+// Temporary auto-lifting. Will be included into DaCe
+LogicalResult liftToPython(TaskletNode &op, JsonEmitter &jemit) {
```

Comment:
This should probably done with pymlir in dace, yes

Reply:
So is it okay if this stays for now? I should maybe add a `NOTE` to expose it to scanners.

Reply:
ok if it stays, but a `TODO(later)` is required :)